modules/manage/pages/cluster-maintenance/partition-recovery.adoc

```diff
@@ -0,0 +1,30 @@
+= Partition recovery
+:description: Recover a single partition using the Admin API.
+
+You can use the Redpanda Admin API to recover a partition that appears to be stuck (for example, does not accept writes). This can arise when the partition replicas have lost glossterm:Raft[] consensus, for instance if brokers in a Raft group fail, preventing the group from reaching a majority and electing a new leader. 
+
+Partition recovery is carried out by promoting the best available replica to leader, allowing the partition to become available for produce and consume. Typically, Redpanda chooses the replica with the highest offset. However, this allows some potential data loss on the partition if the best available replica is out of sync, resulting in an unclean leader election. Partition recovery should only be used when brokers have failed beyond recovery to the point that the remaining replicas cannot form a majority. 
+
+If you want to instead permanently remove failed brokers from a cluster, especially in situations where no surviving partition replicas remain, use nodewise recovery.  
+
+== Use the Admin API to recover a partition
+
+Make a POST request to the `/force_replicas` endpoint:
+
+[,bash]
+----
+curl -X POST http://localhost:9644/v1/debug/partitions/<namespace>/<topic>/<partition>/force_replicas \
+  -d '{
+    [
+        { "core": 0, "node_id": 0 } 
+    ]
+  }'
```

Comment:
The request body schema requires an array of replica assignments. What is the definition of `core` and how do we find its values?

Reply:
core is the CPU core aka shard id. If there are n CPU cores on the machine, user has to pick a value in range [0, n-1], currently user has to pick a random value or find a least loaded shard (if they know what that is) and use that value.

Reply:
@kbatuigas @bharathv this doesn't talk about specifying the actual partition ID which seems to be the most important thing!

I get that you have to specify the node and core (which seems awkward in itself), but what about the specific partitions.  We need to give an example with the actual partition ID you're using.

Reply:
@mattschumpert Looks like the partition ID is specified in the request URL `/debug/partitions/{namespace}/{topic}/{partition}/force_replicas`. @bharathv can you please provide example values for the following?
- `namespace`
- `topic`
- `partition` 
- `node_id`

Reply:
Right, partition ID is in the URL,


Here is a working example, lets assume partition test/0 is leaderless because majority of replicas are gone. 

A user is expected to get the current node/shard assignments using the following command

```
curl http://$(hostname):9644/v1/partitions/kafka/test | jq

[
  {
    "ns": "kafka",
    "topic": "test",
    "partition_id": 0,
    "status": "done",
    "leader_id": 1,
    "raft_group_id": 1,
    "replicas": [
      {
        "node_id": 1,
        "core": 1
      },
      {
        "node_id": 3,
        "core": 1
      },
      {
        "node_id": 2,
        "core": 1
      }
    ]
  }
]
```


Lets assume nodes 2, 3 are dead and we want to replace them with nodes 4, 5, the command would look like this.


```
curl -X POST http://$(hostname):9644/v1/partitions/kafka/test/0/force_replicas  -H 'Content-Type: application/json' --data '[{"node_id": 1, "core": 1}, {"node_id": 4, "core": 1}, {"node_id": 5, "core": 0}]'
```

As you can see the input to the command is the new replica set that will force replace the current one. It includes {1, 1} (which ensures we recover data from it) and two new replica assignments {4, 1}, {5, 0} which will hydrate data from {1, 1}.

User has to pick least loaded cores and the system may later move around the replicas if they are imbalanced.

Clearly this a low level operation and has to be done with caution.